src/deploy/functions/triggerRegionHelper.ts

```diff
@@ -2,50 +2,72 @@ import * as backend from "./backend";
 };
 
 /**
- * Sets the trigger region to what we currently have deployed
+ * Ensures the trigger regions are set and correct
  * @param want the list of function specs we want to deploy
  * @param have the list of function specs we have deployed
  */
-export async function lookupMissingTriggerRegions(want: backend.Backend): Promise<void> {
+export async function ensureTriggerRegions(want: backend.Backend): Promise<void> {
   const regionLookups: Array<Promise<void>> = [];
   for (const ep of backend.allEndpoints(want)) {
-    if (ep.platform === "gcfv1" || !backend.isEventTriggered(ep) || ep.eventTrigger.region) {
+    if (ep.platform === "gcfv1" || !backend.isEventTriggered(ep)) {
       continue;
     }
-    const lookup = LOOKUP_BY_EVENT_TYPE[ep.eventTrigger.eventType];
+    const lookup = ENSURE_BY_EVENT_TYPE[ep.eventTrigger.eventType];
     if (!lookup) {
       logger.debug(
-        "Don't know how to look up trigger region for event type",
+        "Don't know how to ensure trigger region for event type",
         ep.eventTrigger.eventType,
         ". Deploy will fail unless this event type is global"
       );
       continue;
     }
-    regionLookups.push(lookup(ep));
+    regionLookups.push(lookup(ep, ep.region));
   }
   await Promise.all(regionLookups);
 }
 
-/** Sets a GCS event trigger's region to the region of its bucket. */
-async function lookupBucketRegion(endpoint: backend.EventTriggered): Promise<void> {
-  try {
-    const bucket: { location: string } = await storage.getBucket(
-      endpoint.eventTrigger.eventFilters.bucket!
-    );
-    endpoint.eventTrigger.region = bucket.location.toLowerCase();
-  } catch (err) {
-    throw new FirebaseError("Can't find the storage bucket region", { original: err });
+/**
+ * Sets a GCS event trigger's region to the region of its bucket if unset,
+ * and checks for an invalid EventArc trigger region before deployment of the function
+ */
+async function ensureStorageTriggerRegion(
+  endpoint: backend.EventTriggered,
+  endpointRegion: string
+): Promise<void> {
+  if (!endpoint.eventTrigger.region) {
+    logger.debug("Looking up bucket region for the storage event trigger");
+    try {
+      const bucket: { location: string } = await storage.getBucket(
+        endpoint.eventTrigger.eventFilters.bucket!
+      );
+      endpoint.eventTrigger.region = bucket.location.toLowerCase();
+      logger.debug("Setting the event trigger region to", endpoint.eventTrigger.region, ".");
+    } catch (err) {
+      throw new FirebaseError("Can't find the storage bucket region", { original: err });
+    }
+  }
+  // check for invalid cloud storage trigger region
+  if (
+    endpointRegion !== endpoint.eventTrigger.region &&
+    endpoint.eventTrigger.region !== "us-central1" &&
+    !regionInLocation(endpointRegion, endpoint.eventTrigger.region!)
+  ) {
+    throw new FirebaseError("Function cannot be deployed outside of the trigger region");
```

Comment:
Our customers probably aren't thinking about "trigger regions" since we don't ask them to provide one. Can we say something like `A function in region ${endpoint.region} cannot listen to a bucket in region ${endpoint.eventTrigger.region}` ?